### PR TITLE
ipatests: add PQC (ML-DSA) install coverage and basic cert checks

### PR DESCRIPTION
**Pull request by 06shalini**
Enhance PQC/ML-DSA Certificate Integration Test Coverage
This PR significantly expands the integration test suite for Post-Quantum Cryptography (PQC) and ML-DSA certificates within FreeIPA. The primary goal is to ensure robust validation of ML-DSA CA behavior, master/client certificate issuance, and certificate enrollment processes, including new tests for replica CA functionality.
## Integration Tests Coverage
The existing test coverage, now updated and expanded, includes:
### 1. ML-DSA CA Behavior (`TestInstallMasterClientMLDSACA`)
This test class validates the core functionality of an ML-DSA enabled Certificate Authority.
- ✅ `test_ca_signing_keys_are_mldsa` - Verifies that the CA correctly uses ML-DSA signing keys.
- ✅ `test_user_cert_rsa_csr_signed_by_mldsa_ca` - Ensures RSA user certificates can be signed by an ML-DSA CA.
- ✅ `test_user_cert_mldsa_csr_signed_by_mldsa_ca` - Validates ML-DSA user certificates issued by an ML-DSA CA.
- ✅ `test_getcert_mldsa_key_signed_by_mldsa_ca` - Tests master host certificate issuance with ML-DSA keys.
- ✅ `test_getcert_rsa_key_signed_by_mldsa_ca` - Tests master host certificate issuance with RSA keys.
- ✅ `test_getcert_mldsa_key_client_signed_by_mldsa_ca` - Validates client host certificate issuance with ML-DSA keys.
- Inherited Tests: Plus 9 inherited tests from `TestInstallMasterClient` covering certificate tracking, profile validation, SAN handling, and SubCA scenarios.
### 2. Master/Client Certificates (`TestInstallMasterClientMLDSA`)
This class focuses on the issuance of master and client certificates with ML-DSA keys against an RSA CA.
- ✅ `test_user_cert_mldsa_csr_signed_by_rsa_ca` - Verifies user certificate requests with ML-DSA are signed by an RSA CA.
- ✅ `test_getcert_mldsa_key_signed_by_rsa_ca` - Ensures host certificate requests with ML-DSA are signed by an RSA CA.
- Inherited Tests: Plus 9 inherited tests from `TestInstallMasterClient`.
### 3. Enrollment Tests (Master + Replica)
These tests cover the certificate enrollment process on both master and replica servers.
- `TestPQCCertEnrollmentIPACerts` (ML-DSA IPA + RSA CA)
- `TestPQCCertEnrollmentCACerts` (ML-DSA IPA + ML-DSA CA)
- ✅ `test_getcert_enroll_caIPAserviceCert_mldsa_key_master` - Validates Certmonger enrollment on the master.
- ✅ `test_getcert_enroll_caIPAserviceCert_mldsa_key_replica` - Validates Certmonger enrollment on the replica.
- ✅ `test_ipa_cert_request_mldsa_caIPAservice_profile_master` - Tests direct `ipa cert-request` on the master.
## New Additions: Replica CA Functionality Test
A significant addition in this PR is the new test case:
- `test_replica_ca_issues_cert_with_mldsa()` in `TestInstallPQCBase`.
This test specifically validates:
1.  **Replica CA Issuance**: Confirms that a replica CA can successfully issue certificates, ensuring CA replication is functional.
2.  **ML-DSA Key Type Validation**: Verifies that certificates issued from the replica possess the correct ML-DSA key type.
3.  **Full Workflow**: Covers the end-to-end process from user creation, CSR generation, certificate request, to final certificate validation.
**Commands Executed within the test:**
```bash
# On replica:
ipa user-add pqc_replica_test_user
openssl genpkey -algorithm ML-DSA-65 -out /tmp/replica_test.key
openssl req -new -key /tmp/replica_test.key -out /tmp/replica_test.csr
ipa cert-request --principal pqc_replica_test_user /tmp/replica_test.csr
openssl x509 -in /tmp/replica_test.crt -noout -text | grep Public-Key


**Tests are executed in Jenkins on fedora-rawhide. with copr enabled : shalini10/freeipa.** 
Metadata ; Working pipeline results : 
https://{gitlab}/-/snippets/12236
https://{ARTIFACE_SERVER}/idm-ci/idm-ci/trigger/prod/trigger/run/null/16024/report.html?sort=result
https://{gitlab}/-/snippets/12233
https://{ARTIFACE_SERVER}/idm-ci/idm-ci/trigger/prod/trigger/run/null/16020/report.html?sort=result

## Summary by Sourcery

Expand FreeIPA’s PQC integration coverage to validate ML-DSA certificate installation, issuance, and enrollment across supported CA and topology configurations.

New Features:
- Add integration coverage for ML-DSA certificate issuance with RSA and ML-DSA CSRs, service certificates, and CA configurations.
- Validate ML-DSA certificate enrollment on masters and replicas, including replica CA issuance.

Enhancements:
- Refactor PQC installation helpers to support optional replicas and validate configured CA key types.

CI:
- Run dedicated PQC integration suites on Fedora Rawhide with master, replica, and client topologies.

Tests:
- Add ML-DSA certificate integration tests covering CA keys, certificate public keys, certmonger enrollment, SAN handling, and multiple ML-DSA parameter sets.

### CODE REVIEW

**Comment by 06shalini** [comment]
My idea of extended PQC integration test is to extend existing TestInstallMasterClient , but currently some functionalities are not supported fully. Tested the complete TestInstallMasterClient with MLDSA support [here: https://github.com/freeipa/freeipa/pull/8404]: 

Following are current observations :

1. LWCA installation failing : ML-DSA Key not supported : https://{JIRA_URL}/browse/IDM-5147
2. ML-DSA 65, 87 CA installation failing , PR exist with fix:  https://github.com/freeipa/freeipa/pull/8399
3. ipa ca-show ipa failing for mldsa cert : 400 Client Error: Bad Request for url:  Created a JIRA ticket : https://{JIRA_URL}/browse/IDM-6497

Because of the failures, currently minimal support is added to extend the test coverage for PQC.

File: ipatests/test_integration/test_installation.py [hunk 1]
```diff
@@ -2271,9 +2271,39 @@ def test_replica_key_sizes(self):
...
+      `openssl x509 -text` output)
+    - Dogtag CA NSS database contains the expected key type for the signing
+      keys (counted via `certutil -K | grep -c`)
+    """
+
     ipa_key_type = None
     ca_key_type = None
 
+    @pytest.fixture(autouse=True)
+    def xfail_large_mldsa_key_sizes(self):
```
Comment (06shalini) [inline ipatests/test_integration/test_installation.py]: This block will be removed once this PR is merged : https://github.com/freeipa/freeipa/pull/8399

File: ipatests/test_integration/test_installation.py [hunk 2]
```diff
@@ -2281,10 +2295,18 @@ def install(cls, mh):
             extra_args.extend(["--key-type-size", cls.ipa_key_type])
         if cls.ca_key_type:
             extra_args.extend(["--ca-key-type", cls.ca_key_type])
-        tasks.install_master(cls.master, setup_dns=True,
-                             extra_args=extra_args)
-        tasks.install_replica(
-            cls.master, cls.replicas[0], setup_ca=True)
+        # https://github.com/freeipa/freeipa/pull/8399
```
Comment (06shalini) [inline ipatests/test_integration/test_installation.py]: Will remove xfail on merge of PR#8399

File: ipatests/test_integration/test_installation.py [hunk 3]
```diff
@@ -2354,17 +2378,65 @@ def test_replica_key_sizes(self):
...
-class TestInstallPQCIPACerts(TestInstallPQCBase):
-
-    num_replicas = 1
-    master_with_dns = True
-    ipa_key_type = "mldsa"
-    ca_key_type = None
-
-
-class TestInstallPQCCACerts(TestInstallPQCBase):
+def _make_pqc_install_class(name, *, ipa_key_type, ca_key_type, doc):
```
Comment (menonsudhir) [inline ipatests/test_integration/test_installation.py]: 1. I think its better to use explicit subclasses which matches other integration tests rather than defining new function i.e _make_pqc_install_class

class TestInstallPQCIPACerts(TestInstallPQCBase):
    """Install with ML-DSA for IPA service keys (default size)."""
    num_replicas = 1
    master_with_dns = True
    ipa_key_type = "mldsa"
    ca_key_type = None


class TestInstallPQCIPACertsMLDSA65(TestInstallPQCBase):
    """Install with ML-DSA-65 for IPA service keys."""
    num_replicas = 1
    master_with_dns = True
    ipa_key_type = "mldsa:65"
    ca_key_type = None


class TestInstallPQCIPACertsMLDSA87(TestInstallPQCBase):
    """Install with ML-DSA-87 for IPA service keys."""
    num_replicas = 1
    master_with_dns = True
    ipa_key_type = "mldsa:87"
    ca_key_type = None


class TestInstallPQCCACerts(TestInstallPQCBase):
    """Install with ML-DSA CA (default size) and ML-DSA-44 IPA keys."""
    num_replicas = 1
    master_with_dns = True
    ipa_key_type = "mldsa:44"
    ca_key_type = "mldsa"


class TestInstallPQCCACertsMLDSA65(TestInstallPQCBase):
    """Install with ML-DSA-65 for both IPA keys and CA keys."""
    num_replicas = 1
    master_with_dns = True
    ipa_key_type = "mldsa:65"
    ca_key_type = "mldsa:65"


class TestInstallPQCCACertsMLDSA87(TestInstallPQCBase):
    """Install with ML-DSA-87 for both IPA keys and CA keys."""
    num_replicas = 1
    master_with_dns = True
    ipa_key_type = "mldsa:87"
    ca_key_type = "mldsa:87"

File: ipatests/test_integration/test_cert_pqc.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            ["getcert", "list", "-f", paths.HTTPD_CERT_FILE]
+        )
+        assert "profile: caIPAserviceCert" in result.stdout_text
+
+        result = self.master.run_command(
+            ["getcert", "list", "-n", "Server-Cert cert-pki-ca"]
+        )
+        assert "profile: caServerCert" in result.stdout_text
+
+    def test_user_cert_request_mldsa_csr(self):
```
Comment (menonsudhir) [inline ipatests/test_integration/test_cert_pqc.py]: For test_user_cert_request_mldsa_csr i think we can use the try/finally statement something like this. Similarly we can use try/finally with test_client_host_cert_request_certmonger function as well, 

def test_user_cert_request_mldsa_csr(self):
    user = "pqcuser1"
    csr = "/tmp/pqcuser1.csr"
    key = "/tmp/pqcuser1.key"
    crt = "/tmp/pqcuser1.crt"

    tasks.kinit_admin(self.master)
    tasks.user_add(self.master, user)

    try:
        # ... openssl keygen, CSR creation, cert-request
        # pytest.xfail() / pytest.fail() may fire here
    finally:
        # Always runs, even after xfail/skip/fail
        tasks.kdestroy_all(self.master)
        self.master.run_command(["rm", "-f", csr, key, crt], raiseonerr=False)

**Comment by menonsudhir** [comment]
Also we can include scenarios for ML-DSA  if those are supported.
1. ipa-server-install with external ca for e.g some of the tests are already in https://github.com/freeipa/freeipa/blob/master/ipatests/test_integration/test_external_ca.py

File: ipatests/test_integration/test_installation.py [hunk 5]
```diff
@@ -2354,17 +2378,65 @@ def test_replica_key_sizes(self):
...
+_PQC_INSTALL_VARIANTS = (
+    # IPA keys are ML-DSA, CA keys are RSA (default)
+    (
+        "TestInstallPQCIPACerts",
+        "Install with ML-DSA for IPA service keys (default size).",
+        "mldsa",
+        None,
+    ),
+    (
+        "TestInstallPQCIPACertsMLDSA65",
```
Comment (pranav210798) [inline ipatests/test_integration/test_installation.py]: mldsa and mldsa:65 currently exercise the same size. In ipaserver/install/certs.py, both default_cert_key_strength['mldsa'] and default_ca_key_strength['mldsa'] are "65", so --key-type-size mldsa is equivalent to --key-type-size mldsa:65.

That makes TestInstallPQCIPACerts and TestInstallPQCIPACertsMLDSA65 redundant, and we pay for two separate jobs for the same coverage.

Please keep only one of them (either the default mldsa or the explicit mldsa:65), and reserve dedicated variants for sizes that actually differ (44 / 65 / 87). Same idea if any CA jobs duplicate mldsa vs mldsa:65.

**Comment by 06shalini** [comment]
> Also we can include scenarios for ML-DSA if those are supported.
> 
>     1. ipa-server-install with external ca for e.g some of the tests are already in https://github.com/freeipa/freeipa/blob/master/ipatests/test_integration/test_external_ca.py

@menonsudhir 
Noted. For future test development

**Comment by pranav210798** [comment]
Hi @06shalini Please re-write commit message, as you mentioned the direct links to artifacts server and jenkins lab. Please use a placeholder var for that and then add the links.

File: ipatests/test_integration/test_installation.py [hunk 6]
```diff
@@ -2355,6 +2357,81 @@ def test_replica_key_sizes(self):
         self.check_key_sizes(self.replicas[0])
         self.check_ca_keys(self.replicas[0])
 
+    def test_replica_ca_issues_cert_with_mldsa(self):
+        """Test that replica CA can issue certificates with ML-DSA config.
+
+        Validates:
+        - Replica CA is functional and can issue certificates
+        - Certificate issued from replica matches configured key type
+        - CA replication works correctly for ML-DSA scenarios
```
Comment (pranav210798) [inline ipatests/test_integration/test_installation.py:2369]: this doesn't actually prove the replica CA issued the cert. cloned CA has the same issuer, so you're only checking that cert-request works from the replica host.

File: ipatests/test_integration/test_installation.py [hunk 7]
```diff
@@ -2355,6 +2357,81 @@ def test_replica_key_sizes(self):
...
+            raise pytest.skip("No replica installed. Skipping")
+        result = tasks.kinit_admin(self.replicas[0], raiseonerr=False)
+        if result.returncode != 0:
+            raise pytest.skip("Replica is not available. Skipping")
+
+        replica = self.replicas[0]
+        expected_key_type = self._get_key_type(self.ipa_key_type)
+
+        # Create a test user on replica
+        test_user = "pqc_replica_test_user"
```
Comment (pranav210798) [inline ipatests/test_integration/test_installation.py]: hardcoded username will collide on reruns. use a random suffix like the enrollment tests.

File: ipatests/test_integration/test_installation.py [hunk 8]
```diff
@@ -2355,6 +2357,81 @@ def test_replica_key_sizes(self):
...
+        tasks.kinit_admin(replica)
+        try:
+            tasks.user_add(replica, test_user)
+
+            # Generate CSR with the same key type as IPA installation
+            csr_file = "/tmp/replica_test.csr"
```
Comment (pranav210798) [inline ipatests/test_integration/test_installation.py]: if user_add fails, finally still uses csr_file and you'll get NameError. move the path assignments above the try. also /tmp for private keys please use paths.OPENSSL_* like the other tests.

File: ipatests/test_integration/test_installation.py [hunk 9]
```diff
@@ -2355,6 +2357,81 @@ def test_replica_key_sizes(self):
...
+
+            # Request certificate from replica CA
+            replica.run_command([
+                "ipa", "cert-request", "--principal", test_user,
+                "--certificate-out", cert_file, csr_file
+            ])
+
+            # Validate issued certificate has expected key type
+            result = replica.run_command(
+                f"openssl x509 -in {cert_file} -noout -text | grep Public-Key"
```
Comment (pranav210798) [inline ipatests/test_integration/test_installation.py:2425]: publicKey only shows the CSR key type. please also grep signature algorithm so we know it was signed by the mldsa ca.

File: ipatests/test_integration/test_installation.py [hunk 10]
```diff
@@ -2365,8 +2442,39 @@ class TestInstallPQCIPACerts(TestInstallPQCBase):
 
 
 class TestInstallPQCCACerts(TestInstallPQCBase):
+    """ML-DSA-44 server keys with ML-DSA-65 (default) CA.
```
Comment (pranav210798) [inline ipatests/test_integration/test_installation.py]: docstring doesn't match the class. this is ipa_key_type=mldsa (65) and ca_key_type=mldsa:44, not the other way around also please drop the "see also 006a" bit.

File: ipatests/test_integration/test_cert_pqc.py [hunk 11]
```diff
@@ -0,0 +1,659 @@
...
+            raise ValueError(key_type)
+        return csr_file, key_file
+
+    def _ipa_user_cert_request(self, user, csr_file, cert_file):
+        self.master.run_command([
+            'ipa', 'cert-request', '--principal', user,
+            '--certificate-out', cert_file, csr_file,
+        ])
+
+    def _assert_user_cert_public_key(self, cert_file, key_type):
```
Comment (pranav210798) [inline ipatests/test_integration/test_cert_pqc.py:201]: this only checks Public-Key. for the *_signed_by_mldsa_ca tests we also need Signature Algorithm, otherwise an rsa ca would still pass.

File: ipatests/test_integration/test_cert_pqc.py [hunk 12]
```diff
@@ -0,0 +1,659 @@
...
+                self.master.run_command(
+                    ['rm', '-f', csr_file, key_file, cert_file],
+                    raiseonerr=False
+                )
+            entry = ldap.get_entry(
+                DN(('uid', user), ('cn', 'users'), ('cn', 'accounts'),
+                   self.master.domain.basedn)
+            )
+            assert len(entry.get('usercertificate')) == len(key_specs)
+        finally:
```
Comment (pranav210798) [inline ipatests/test_integration/test_cert_pqc.py:233]: please user-del in finally as well. and the csr/key files are relative to cwd use paths.OPENSSL_* .

File: ipatests/test_integration/test_cert_pqc.py [hunk 13]
```diff
@@ -0,0 +1,659 @@
...
+            ['getcert', 'stop-tracking', '-i', req_id], raiseonerr=False)
+        host.run_command(
+            ['rm', '-f',
+             os.path.join(paths.OPENSSL_CERTS_DIR, '%s.pem' % req_id),
+             os.path.join(paths.OPENSSL_PRIVATE_DIR, '%s.key' % req_id)],
+            raiseonerr=False,
+        )
+
+
+class TestInstallMasterClientMLDSA(PQCInstallBase,
```
Comment (pranav210798) [inline ipatests/test_integration/test_cert_pqc.py:279]: this still inherits test_getcert_list_profile_using_subca and test_remove_missing_lwca. those hit LWCA (IDM-5147) and ca-show (IDM-6497). skip/xfail them here, same as the other inherited tests you already skipped.

File: ipatests/test_integration/test_cert_pqc.py [hunk 14]
```diff
@@ -0,0 +1,659 @@
...
+
+    def test_ca_signing_keys_are_mldsa(self):
+        """Dogtag CA signing keys use ML-DSA."""
+        result = self.master.run_command(
+            'certutil -K -d %s -f %s | grep -c mldsa' % (
+                paths.PKI_TOMCAT_ALIAS_DIR,
+                paths.PKI_TOMCAT_ALIAS_PWDFILE_TXT,
+            )
+        )
+        assert '5' in result.stdout_text
```
Comment (pranav210798) [inline ipatests/test_integration/test_cert_pqc.py]: this will also pass on 15. please parse it as int and assert == 5, same as check_ca_keys in test_installation.py.

**Review by pranav210798 — COMMENTED** [review]
Hi @06shalini Please find my inline comments, Free to discuss.

File: ipatests/test_integration/test_cert_pqc.py (hunk 13 above)
Comment (06shalini) [reply ipatests/test_integration/test_cert_pqc.py:279]: I think both of these issues should be fixed, And my tests are passing. I added the detailed findings in the tickets.

File: ipatests/test_integration/test_cert_pqc.py (hunk 11 above)
Comment (06shalini) [reply ipatests/test_integration/test_cert_pqc.py:201]: Which is a valid use case. RSA CA should be able to sign ML-DSA cert.

**Comment by 06shalini** [comment]
> Hi @06shalini Please re-write commit message, as you mentioned the direct links to artifacts server and jenkins lab. Please use a placeholder var for that and then add the links.

Will club all the commits into one before merge.

**Comment by agaragna77** [comment]
Hi @06shalini -- reviewed this after our Slack chat last week. 
This looks good to merge as-is. 

Capturing a few notes here from our review so they don't live in Slack only:

getcert request usage: Your -G ML-DSA-65 approach works fine here. When an existing key file is passed with -k, certmonger can ignore -G and use the key as-is, so you don't need -g alongside explicit -G. If the command yields a valid cert, that's enough to prove IPA/Dogtag handles ML-DSA enrollment across these paths.

Follow-ups for a second PR (not blocking this one):

Rekeying on TestInstallMasterClientMLDSACA: same-size (65->65) and cross-size (65->87). The cross-size test will need the updated certmonger package (PR #315 changes aren't in rawhide/COPR yet).

Resubmit test: confirm the key is preserved for PQC certs.

Explicit ML-DSA rekey tests: don't rely on inheritance from RSA rekey tests; PQC needs explicit coverage.

Certmonger dependency: tests are running on certmonger 0.79.21-9.fc45 right now. Cross-strength rekeying and edge cases need the newer build to land in rawhide/COPR.

Out of scope for now: LWCA ML-DSA install, ML-KEM, and external CA scenarios make sense to hold for follow-up PRs once underlying support is in place.

I saw the comments from @menonsudhir and @pranav210798 -- agreed that external CA scenarios fit better in a follow-up, and I'm fine with Shalini cleaning up the commit message right before merging.

LGTM to merge as the first PQC baseline. I can re-review the rekey work once the updated certmonger package lands.

**Comment by flo-renaud** [comment]
Hi @06shalini 

I finally fixed the issues in PRCI with Rawhide. Please rebase your PR on top of the master branch, and update your temp commit with **version: 0.10.2** and the tests should be able to build and run.
